### PR TITLE
Product

### PR DESCRIPTION
**Pull request by FatmaQunnies24**
I implemented the Product Class, which includes the essential properties such as id, name, description, price, stock, and isAvailable. Additionally, CRUD operations were implemented.

### CODE REVIEW

**Review by ayanourr — APPROVED** [review]

File: app/api/routes/product_routes.py
Comment (Mohammed-samaheen) [inline app/api/routes/product_routes.py:1]: Please follow the same convention; no need to add `route` to the file name when you are in the route folder

File: app/api/routes/product_routes.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        "price": 150.0,
+        "created_at": datetime.now(),
+        "updated_at": datetime.now(),
+    },
+}
+
+# Create product (only admin)
+@router.post("/products", response_model=Product)
+async def create_product(product: CreateProductRequest, current_admin: User = Depends(get_current_admin)):
+    new_product = Product(**product.dict(), created_at=datetime.now(), updated_at=datetime.now())
```
Comment (Mohammed-samaheen) [inline app/api/routes/product_routes.py]: Update date for a new product?

File: app/api/routes/product_routes.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    products_db[new_product.id] = new_product.dict()
```
Comment (Mohammed-samaheen) [inline app/api/routes/product_routes.py]: We need to verify if the product already exists. Please take a look at the ticket for all the details.

File: app/api/routes/product_routes.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    product['updated_at'] = datetime.now()  
+    return product
+
+# Delete product by ID (only admin )
+@router.delete("/products/{product_id}", response_model=Product)
+async def delete_product(product_id: UUID, current_admin: User = Depends(get_current_admin)):
+    product = products_db.pop(product_id, None)
+    if product is None:
+        raise HTTPException(status_code=404, detail="Product not found")
+    return product
```
Comment (Mohammed-samaheen) [inline app/api/routes/product_routes.py]: in the delete method, we return nothing. If the deletion was successful, we only return a status code `status.HTTP_204_NO_CONTENT`

File: app/api/routes/product_routes.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+# Update product by ID (only admin)
+@router.put("/products/{product_id}", response_model=Product)
+async def update_product(product_id: UUID, updated_data: UpdateProductRequest, current_admin: User = Depends(get_current_admin)):
+    product = products_db.get(product_id)
+    if product is None:
+        raise HTTPException(status_code=404, detail="Product not found")
+
+    update_data = updated_data.dict(exclude_unset=True)
+    for key, value in update_data.items():
```
Comment (Mohammed-samaheen) [inline app/api/routes/product_routes.py]: When updating the name, make sure the new name is not already in the database.

Comment (Mohammed-samaheen) [reply app/api/routes/product_routes.py]: You can leave a #TODO comment if you want. When you add the DB setup, you can implement that part.

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

File: app/api/routes/product_routes.py (hunk 1 above)
Comment (FatmaQunnies24) [reply app/api/routes/product_routes.py]: I set the updated_at date to reflect the last time a modification was made. During the creation of a new product, the updated_at date is set to the same value as created_at, which means it takes the current time as the creation date

Comment (Mohammed-samaheen) [reply app/api/routes/product_routes.py]: Yes, that's why it's invalid. The update date is the last update date, and it should only be changed in the update endpoint.

**Comment by Mohammed-samaheen** [comment]
resolve the conflict

**Comment by Mohammed-samaheen** [comment]
Please resolve the conflict. Also, when you make the changes, click on Re-request review; this will notify me to take a look at the PR.
Screenshot 2024-10-16 at 7 57 20 PM

File: requirments.txt
Comment (Mohammed-samaheen) [inline requirments.txt:1]: You accidentally deleted this file

Comment (FatmaQunnies24) [reply requirments.txt:1]: I didn't resolve the conflict; I just tried to resolve it from GitHub.

**Comment by Mohammed-samaheen** [comment]
Please close this PR if you don't need it anymore

**Pull request merged**